File: springcloud/order-service/src/main/java/com/example/orderservice/controller/OrderController.java
package com.example.orderservice.controller;

import com.example.orderservice.dto.OrderDto;
import com.example.orderservice.entity.OrderEntity;
import com.example.orderservice.service.OrderService;
import com.example.orderservice.vo.RequestOrder;
import com.example.orderservice.vo.ResponseOrder;
import lombok.RequiredArgsConstructor;
import org.modelmapper.ModelMapper;
import org.modelmapper.convention.MatchingStrategies;
import org.springframework.core.env.Environment;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import javax.annotation.PostConstruct;
import java.util.ArrayList;
import java.util.List;

@RestController
@RequestMapping("/order-service")
@RequiredArgsConstructor
public class OrderController {

    private final Environment env;
    private final OrderService orderService;
    private final ModelMapper mapper = new ModelMapper();

    @PostConstruct
    public void setMapper() {
        mapper.getConfiguration().setMatchingStrategy(MatchingStrategies.STRICT);
    }
    @PostMapping("/{userId}/orders")
    public ResponseEntity<ResponseOrder> createOrder(@PathVariable("userId") String userId,
                                                     @RequestBody RequestOrder requestOrder) {

        OrderDto orderDto = mapper.map(requestOrder, OrderDto.class);
        orderDto.setUserId(userId);
        OrderDto createdOrder = orderService.createOrder(orderDto);

        ResponseOrder responseOrder = mapper.map(createdOrder, ResponseOrder.class);

        return ResponseEntity.status(HttpStatus.CREATED).body(responseOrder);
    }

    @GetMapping("/{userId}/orders")
    public ResponseEntity<List<ResponseOrder>> getOrders(@PathVariable("userId") String userId) {
        List<OrderEntity> orders = orderService.getOrdersByUserId(userId);
        List<ResponseOrder> result = new ArrayList<>();

        orders.forEach(orderEntity -> result.add(mapper.map(orderEntity, ResponseOrder.class)));

        return ResponseEntity.status(HttpStatus.OK).body(result);
    }
}

File: distributed-transaction/saga-choreography-order/src/main/kotlin/org/example/order/controller/OrderController.kt
package org.example.order.controller

import org.example.order.application.OrderService
import org.example.order.application.RedisLockService
import org.example.order.controller.dto.CreateOrderRequest
import org.example.order.controller.dto.CreateOrderResponse
import org.example.order.controller.dto.PlaceOrderRequest
import org.example.order.domain.OrderStatus
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController

@RestController
class OrderController(
    private val orderService: OrderService,
    private val redisLockService: RedisLockService,
) {

    @PostMapping("/orders")
    fun createOrder(
        @RequestBody request: CreateOrderRequest
    ): CreateOrderResponse {
        val result = orderService.createOrder(request.toCommand())
        return CreateOrderResponse(result.orderId)
    }

    @PostMapping("/orders/place")
    fun placeOrder(
        @RequestBody request: PlaceOrderRequest,
    ) {
        val key = "order:choreography:${request.orderId}"

        val acquiredLock = redisLockService.tryLock(key, request.orderId.toString())

        if (!acquiredLock) {
            throw RuntimeException("락 획득에 실패하였습니다.")
        }

        try {
            orderService.placeOrder(request.toCommand())
        } finally {
            redisLockService.releaseLock(key)
        }
    }

    @GetMapping("/orders/{orderId}/status")
    fun getStatus(
        @PathVariable orderId: Long,
    ): OrderStatus {
        return orderService.getStatus(orderId)
    }

}